url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/tffst/Tffst.java

```diff
@@ -0,0 +1,1712 @@
+              if (state != initial) {
+                for (Transition inTrans : state.getArrivingTransitions()) {
+                  State origin = inTrans.getFrom();
+                  if (origin != null) {
+                    if (!state.equals(origin)) {
+                      TfString lbout = (inTrans.labelOut.clone());
+                      lbout.addAll(t.labelOut);
+                      new Transition(origin, inTrans.labelIn.clone(),lbout , t.getTo());
+                      workingStates.add(origin);
+                    } else {
+                      if (!inTrans.labelIn.isEpsilon()) {
+                        State n = new State();
+                        new Transition(n, inTrans.labelIn, inTrans.labelOut, n);
+                        workingStates.add(n);
+                        for (Transition arriTransition2 : state.getArrivingTransitions()) {
+                          State origin2 = arriTransition2.getFrom();
+                          if (!state.equals(origin2)) {
+                            new Transition(origin2, arriTransition2.labelIn, inTrans.labelOut, n);
+                            workingStates.add(origin2);
+                          }
+                        }
+                        new Transition(n, inTrans.labelIn, t.labelOut, t.getTo());
+                      }
+                    }
+                  }
+                }
+              } else {
+                for (Transition tNext : t.getTo().getTransitions()) {
+                  if (!(tNext.labelIn.isEpsilon() && tNext.getTo().equals(tNext.getFrom()))) {
+                    TfString lbout = tNext.labelOut.clone();
+                    for (int i = 0; i < t.labelOut.size(); i++)
+                      lbout.add(i, t.labelOut.get(i));
+                    Transition tNextClone = new Transition(null, tNext.labelIn.clone(), lbout, tNext.getTo());
+                    toAdd.add(tNextClone);
+                    removeState = false;
+                  }
+                }
+              }
+            }
+          }
+        }
+        state.removeAllTransitions(toRemove);
+        state.addAllTransitions(toAdd);
+        if (removeState) workingStates.remove(state);
+      }
+    } catch (CloneNotSupportedException e) {
+      e.printStackTrace();
+    }
+    
+    removeDeadTransitions();
+  }
+
+  
+  
+  /**
+   * returns a Tffsr representing this tffst. it expands subEpochs in simpe
+   * pairts of TFs adding epsilons
+   * 
+   * @return
+   */
+  public Tffsr firstProjection() {
+   
+    Tffsr outTffsr = new Tffsr();
+    Map<State, uy.edu.fing.mina.fsa.tffsr.State> stateMap = new HashMap<State, uy.edu.fing.mina.fsa.tffsr.State>();
+
+    for (State fstState : getStates()) {
+      uy.edu.fing.mina.fsa.tffsr.State fsrState = new uy.edu.fing.mina.fsa.tffsr.State();
+      fsrState.setAccept(fstState.isAccept());
+      if (initial.equals(fstState)) 
+        outTffsr.initial = fsrState;
+      stateMap.put(fstState, fsrState);
+    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 690:696, 1242:1248.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.